### ISSUE TITLE
Inject Touch events like Tap/Scroll

### ISSUE THREAD

**Issue by rohitsangwan01** [danielpaulus/go-ios#404]
Thankyou for this awesome project, 
it would be really great if we can somehow inject touch events or Mouse events
or if we can programatically simulate a mouse, not sure if this fits in scope of go-ios

**Comment by gotoin** [comment]
+1

**Comment by nanoscopic** [comment]
This has been asked before. There is a ui event synthesis system through rsd that is used by xcode for some things ( such as the "dump ui structure" thing in xcode. That theoretically could do it.

No one has figured it out yet though, at least not publicly.

I've managed to write code that can do events through the mechanism universal control uses, but that only works for ipads.

**Comment by ida-lovelace** [comment]
Has anyone figured this out?

**Comment by nanoscopic** [comment]
I've extensively explained this over, and over, and over again.

It can be done via:

- XCTest ( Synthetic events ) WDA etc
- Universal Control
- Private Apple only automation ( uses the same underlying stuff Universal Control does, is gated and only Apple internal accessible ) I didn't mention this before because it is restricted and unavailable despite existing on the devices.
- Bluetooth mouse
- USB mouse ( relative movement only )
- USB touchpad ( relative movement only )
- USB keyboard ( via mousekeys )

I've dug into all of these. Unless Apple adds something additional and new to a new iOS version, there is nothing more to know.

What should be investigated and fully reimplemented via RE is everything related to Continuity, which is what Universal Control is built on top of. It looks like there -may- be additional mechanisms in there connected to this stuff. Example being I'm not entirely sure Universal Control isn't available in a sense on iphones despite that Universal Control the official feature only works on ipads.

Nobody though is working on that publicly, and nobody seems willing to fund the work to figure it out. So long as people as just asking for this stuff for free and unwilling to pay those who know how to do this, it is not going to happen.

**Comment by ida-lovelace** [comment]
> I've extensively explained this over, and over, and over again.
> 
> It can be done via:
> 
> - XCTest ( Synthetic events ) WDA etc
> - Universal Control
> - Private Apple only automation ( uses the same underlying stuff Universal Control does, is gated and only Apple internal accessible ) I didn't mention this before because it is restricted and unavailable despite existing on the devices.
> - Bluetooth mouse
> - USB mouse ( relative movement only )
> - USB touchpad ( relative movement only )
> - USB keyboard ( via mousekeys )
> 
> I've dug into all of these. Unless Apple adds something additional and new to a new iOS version, there is nothing more to know.
> 
> What should be investigated and fully reimplemented via RE is everything related to Continuity, which is what Universal Control is built on top of. It looks like there -may- be additional mechanisms in there connected to this stuff. Example being I'm not entirely sure Universal Control isn't available in a sense on iphones despite that Universal Control the official feature only works on ipads.
> 
> Nobody though is working on that publicly, and nobody seems willing to fund the work to figure it out. So long as people as just asking for this stuff for free and unwilling to pay those who know how to do this, it is not going to happen.

There's a piece of software called Wormhole that manages to do this, although the dev doesn't have plans to open source it, it's a demo (otherwise paid) and Windows only; it uses a hacky Bluetooth HID emulation trick and makes use of iOS touchassist. It's a little difficult to get paired at first, but it does get the job done.

Leaving it here in case someone gets inspired and has the knowledge to code public, cross-platform builds: https://er.run/

**Comment by nanoscopic** [comment]
Yes. And? That's what I'm referring to with "Bluetooth mouse" here.

The source for that shit isn't needed, because there are tons of Bluetooth device emulator projects on GitHub, including ones targeted at the relative movement that iOS supports. Which is all it supports. It does not support absolute coordinate bluetooth mice, so it is effectively junk. This is, though, how many iOS device farms work.

You are adding literally nothing to the conversation that people working on this stuff don't already know.

If you want to fucking inspire, pay someone or donate some money to "inspire" them, because you're post ain't doing it for me or anyone else.

**Comment by ida-lovelace** [comment]
Your use of profanity is unnecessary. I added the final part as a note for anyone willing to work on it as a passion project, and to make the code public, that's all. I wasn't talking about you specifically.

Furthermore, relative positioning is more than enough for me. I'm just looking for a way to use an iPhone remotely, be it for personal use or remote assistance.

I'd love to see some repositories implementing Bluetooth HID emulation properly, the one i've tried (https://github.com/Alkaid-Benetnash/EmuBTHID) seems to not be compatible with Apple devices. 

Thank you.

**Comment by nanoscopic** [comment]
My profanity is warranted in my view. It is also permitted as long as I am not using it to attack anyone. I am upset about this stuff and continue to be upset about this stuff long term. I've done nothing to hide this and will not begin to hide my opinion and feelings.

You may feel insulted. That isn't optimal but you should know you are wasting my time and the time of others who might actually push this stuff forward.

There are plenty of projects I have found over the years that do what you are asking for. I just authored another one with AI yesterday for giggles and I may share it with you. I have zero incentive to do so though.

If I had the links to the projects to do this I would probably share them with you. I don't though I'd have to go searching for them just as you can.

You -cannot- emulate mice with only Apple hardware ( at least not their current laptops or desktops ). I've looked into it. If someone discovers how, great.

You -can- do it with nrf.... boards, that's the most effective way. I own a bunch of them. They are sitting on the shelf behind me. You can also do it with raspberry pis or esp32s.

I sell an entire effective system for controlling iOS devices. I'm well known for this. I'm disincentivized to give away my research and methods at this point because people are not paying for it. If I can't survive on the research I've done for years I'd rather let it die than give it away.

I know what you want. I know how to do it. The limit to what I'm willing to do is explain how. I've done this. I'm doing it now.

Sure maybe someone will discover a way eventually that does what I, you, and everyone wants, which is to be able to control iOS devices without all the Apple signing, paid developer accounts, etc etc. I'm telling you as one of the people most knowledgable about this, it's unlikely to happen, and definitely not by well wishers.

**Comment by danielpaulus** [comment]
Summarizing where this thread landed: there is no public, non-XCTest touch-injection mechanism on real iOS devices — the CoreDevice/RSD event-synthesis path is Apple-internal, and the HID routes (Bluetooth/USB mouse emulation) need external hardware and are outside go-ios's scope. The supported path with go-ios is WebDriverAgent via `ios runwda`, which exposes tap/swipe/typing over HTTP. Closing since there's nothing actionable in go-ios itself; if Apple ever exposes a usable service, a fresh issue with details is welcome.

_(part of an automated backlog sweep — comment or reopen if this got it wrong)_
